### DISCUSSION TITLE
Pyinstaller with Nornir -  plugin 'netmiko' is not registered

### DISCUSSION THREAD

**Discussion by baddington1667** [nornir-automation/nornir#668] [Q&A] [answered]
Trying to package a Nornir script with pyinstaller I am receiving the error. Any help on how to resolve this? It is worth noting, when not packaging with pyinstaller the script works.

nornir.core.exceptions.PluginNotRegistered: plugin 'netmiko' is not registered

* CSR8 ** changed : False ******************************************************
vvvv snmpUpdate ** changed : False vvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvv ERROR
Subtask: SNMP Update (failed)

---- SNMP Update ** changed : False -------------------------------------------- ERROR
Traceback (most recent call last):
  File "nornir/core/task.py", line 98, in start
  File "nornir_netmiko/tasks/netmiko_send_config.py", line 24, in netmiko_send_config
  File "nornir/core/inventory.py", line 440, in get_connection
  File "nornir/core/inventory.py", line 481, in open_connection
  File "nornir/core/plugins/register.py", line 75, in get_plugin
nornir.core.exceptions.PluginNotRegistered: plugin 'netmiko' is not registered

^^^^ END snmpUpdate ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

main.spec:

# -*- mode: python ; coding: utf-8 -*-

block_cipher = None

a = Analysis(['main.py'],
             pathex=['/Users/bradleyaddington/Repositories/NetworkAutomation'],
             binaries=[],
             datas=[('config.yml', '.'), ('configuration_files/auvikNetflowConfigG00_10.txt', '.'), ('configuration_files/auvikNetflowConfigGi000_10.txt', '.'), ('configuration_files/auvikNetflowConfigP1_10.txt', '.'), ('configuration_files/auvikNetflowConfigP2_10.txt', '.'), ('inventory/defaults.yml', '.'), ('inventory/groups.yml', '.'), ('inventory/hosts.yml', '.')],
             hiddenimports=[],
             hookspath=['hooks'],
             runtime_hooks=[],
             excludes=[],
             win_no_prefer_redirects=False,
             win_private_assemblies=False,
             cipher=block_cipher,
             noarchive=False)
pyz = PYZ(a.pure, a.zipped_data,
             cipher=block_cipher)
exe = EXE(pyz,
          a.scripts,
          a.binaries,
          a.zipfiles,
          a.datas,
          [],
          name='main',
          debug=False,
          bootloader_ignore_signals=False,
          strip=False,
          upx=True,
          upx_exclude=[],
          runtime_tmpdir=None,
          console=True )

**Comment by dmfigol** [comment, marked as answer]
Please format your code in [markdown code blocks](https://guides.github.com/features/mastering-markdown/#examples).  
I haven't used PyInstaller, but I suspect when it packages the application, it doesn't bundles plugins to the appropriate entrypoints, so this code doesn't register anything:
https://github.com/nornir-automation/nornir/blob/e4f6b8c6258ae2dcfb286098b30652c7a31ecf30/nornir/core/plugins/register.py#L19-L25
I would suggest to try registering the plugin manually in your code, e.g.:
```
from nornir.core.plugins.connections import ConnectionPluginRegister
from nornir_netmiko.connections import Netmiko

ConnectionPluginRegister.register("netmiko", Netmiko)
```

**Reply by baddington1667** [reply]
That worked! Thank you and also thank you for the link on markdown code blocks. I am relatively new to coding and GitHub so still learning.
